File: src/main/java/com/refinedmods/refinedstorage/item/SecurityCardItem.java
package com.refinedmods.refinedstorage.item;

import com.refinedmods.refinedstorage.RS;
import com.refinedmods.refinedstorage.api.network.security.Permission;
import com.refinedmods.refinedstorage.render.Styles;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.TextComponent;
import net.minecraft.network.chat.TranslatableComponent;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResultHolder;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.TooltipFlag;
import net.minecraft.world.level.Level;

import javax.annotation.Nullable;
import java.util.List;
import java.util.UUID;

public class SecurityCardItem extends Item {
    private static final String NBT_OWNER = "Owner";
    private static final String NBT_OWNER_NAME = "OwnerName";
    private static final String NBT_PERMISSION = "Permission_%d";

    public SecurityCardItem() {
        super(new Item.Properties().tab(RS.CREATIVE_TAB).stacksTo(1));
    }

    @Nullable
    public static UUID getOwner(ItemStack stack) {
        if (stack.hasTag() && stack.getTag().contains(NBT_OWNER)) {
            return UUID.fromString(stack.getTag().getString(NBT_OWNER));
        }

        return null;
    }

    public static boolean hasPermission(ItemStack stack, Permission permission) {
        String id = String.format(NBT_PERMISSION, permission.getId());

        if (stack.hasTag() && stack.getTag().contains(id)) {
            return stack.getTag().getBoolean(id);
        }

        return false;
    }

    public static void setPermission(ItemStack stack, Permission permission, boolean state) {
        if (!stack.hasTag()) {
            stack.setTag(new CompoundTag());
        }

        stack.getTag().putBoolean(String.format(NBT_PERMISSION, permission.getId()), state);
    }

    public static boolean isValid(ItemStack stack) {
        return stack.hasTag() && stack.getTag().contains(NBT_OWNER);
    }

    @Override
    public InteractionResultHolder<ItemStack> use(Level level, Player player, InteractionHand hand) {
        ItemStack stack = player.getItemInHand(hand);

        if (!level.isClientSide) {
            stack.setTag(new CompoundTag());

            stack.getTag().putString(NBT_OWNER, player.getGameProfile().getId().toString());
            stack.getTag().putString(NBT_OWNER_NAME, player.getGameProfile().getName());
        }

        return InteractionResultHolder.success(stack);
    }

    @Override
    public void appendHoverText(ItemStack stack, @Nullable Level level, List<Component> tooltip, TooltipFlag flag) {
        super.appendHoverText(stack, level, tooltip, flag);

        if (stack.hasTag() && stack.getTag().contains(NBT_OWNER_NAME)) {
            tooltip.add(new TranslatableComponent("item.refinedstorage.security_card.owner", stack.getTag().getString(NBT_OWNER_NAME)).setStyle(Styles.GRAY));
        }

        for (Permission permission : Permission.values()) {
            if (hasPermission(stack, permission)) {
                tooltip.add(new TextComponent("- ").append(new TranslatableComponent("gui.refinedstorage.security_manager.permission." + permission.getId())).setStyle(Styles.GRAY));
            }
        }
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/item/SecurityCardItem.kt
package com.refinedmods.refinedstorage.item

import com.refinedmods.refinedstorage.RS
import com.refinedmods.refinedstorage.api.network.security.Permission
import com.refinedmods.refinedstorage.render.Styles
import net.minecraft.client.util.ITooltipFlag
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.nbt.CompoundTag
import net.minecraft.util.ActionResult
import net.minecraft.util.Hand
import net.minecraft.util.text.Text
import net.minecraft.util.text.StringTextComponent
import net.minecraft.util.text.TranslationTextComponent
import net.minecraft.world.World
import java.util.*

class SecurityCardItem : Item(Properties().group(RS.MAIN_GROUP).maxStackSize(1)) {
    fun onItemRightClick(world: World, player: PlayerEntity, hand: Hand?): ActionResult<ItemStack> {
        val stack: ItemStack = player.getHeldItem(hand)
        if (!world.isClient) {
            stack.tag = CompoundTag()
            stack.tag!!.putString(NBT_OWNER, player.gameProfile.id.toString())
            stack.tag!!.putString(NBT_OWNER_NAME, player.gameProfile.name)
        }
        return ActionResult.resultSuccess(stack)
    }

    fun addInformation(stack: ItemStack, @Nullable world: World?, tooltip: MutableList<Text?>, flag: ITooltipFlag?) {
        super.addInformation(stack, world, tooltip, flag)
        if (stack.hasTag() && stack.tag!!.contains(NBT_OWNER_NAME)) {
            tooltip.add(TranslationTextComponent("item.refinedstorage.security_card.owner", stack.tag!!.getString(NBT_OWNER_NAME)).setStyle(Styles.GRAY))
        }
        for (permission in Permission.values()) {
            if (hasPermission(stack, permission)) {
                tooltip.add(StringTextComponent("- ").append(TranslationTextComponent("gui.refinedstorage.security_manager.permission." + permission.id)).setStyle(Styles.GRAY))
            }
        }
    }

    companion object {
        private const val NBT_OWNER = "Owner"
        private const val NBT_OWNER_NAME = "OwnerName"
        private const val NBT_PERMISSION = "Permission_%d"
        @Nullable
        fun getOwner(stack: ItemStack): UUID? {
            return if (stack.hasTag() && stack.tag!!.contains(NBT_OWNER)) {
                UUID.fromString(stack.tag!!.getString(NBT_OWNER))
            } else null
        }

        @kotlin.jvm.JvmStatic
        fun hasPermission(stack: ItemStack, permission: Permission): Boolean {
            val id = String.format(NBT_PERMISSION, permission.id)
            return if (stack.hasTag() && stack.tag!!.contains(id)) {
                stack.tag!!.getBoolean(id)
            } else false
        }

        fun setPermission(stack: ItemStack, permission: Permission, state: Boolean) {
            if (!stack.hasTag()) {
                stack.tag = CompoundTag()
            }
            stack.tag!!.putBoolean(String.format(NBT_PERMISSION, permission.id), state)
        }

        fun isValid(stack: ItemStack): Boolean {
            return stack.hasTag() && stack.tag!!.contains(NBT_OWNER)
        }
    }

    init {
        this.setRegistryName(RS.ID, "security_card")
    }
}